doc/user/content/sql/create-sink/kafka.md

```diff
@@ -69,33 +70,252 @@ Field                | Value  | Description
+  ("col2")
+  ("col3", "col4")
+```
+
+The first error message indicates that Materialize could not prove the existence
+of any unique keys for the sink's source relation. The second error message
+indicates that Materialize could prove that `col2` and `(col3, col4)` were
+unique keys of the sink's source relation, but could not provide the uniqueness
+of the specified upsert key of `col1`.
+
+There are three ways to resolve this error:
+
+* Change the sink to use one of the keys that Materialize determined to be
+  unique, if such a key exists and has the appropriate semantics for your
+  use case.
+
+* Create a materialized view that deduplicates the input relation by the
+  desired upsert key:
+
+  ```sql
+  -- For each row with the same key `k`, the `ORDER BY` clause ensures we
+  -- keep the row with the largest value of `v`.
+  CREATE MATERIALIZED VIEW deduped AS
+  SELECT DISTINCT ON (k) v
+  FROM original_input
+  ORDER BY k, v DESC;
+
+  -- Materialize can now prove that `k` is a unique key of `deduped`.
+  CREATE SINK s IN CLUSTER my_io_cluster
+  FROM deduped
+  INTO KAFKA CONNECTION kafka_connection (TOPIC 't')
+  KEY (k)
+  FORMAT JSON ENVELOPE UPSERT;
+  ```
+
+  {{< note >}}
+  Maintaining the `deduped` materialized view requires memory proportional to the
+  number of records in `original_input`. Be sure to assign `deduped`
+  to a cluster with resources sufficient to handle your data volume.
+  {{< /note >}}
+
+* Use the `NOT ENFORCED` clause to disable Materialize's validation of the key's
+  uniqueness:
+
+  ```sql
+  CREATE SINK s IN CLUSTER my_io_cluster
+  FROM original_input
+  INTO KAFKA CONNECTION kafka_connection (TOPIC 't')
+  -- We have outside knowledge that `k` is a unique key of `original_input`, but
+  -- Materialize cannot prove this, so we disable its key uniqueness check.
+  KEY k NOT ENFORCED
+  FORMAT JSON ENVELOPE UPSERT;
+  ```
+
+  You should only disable this verification if you have outside knowledge of
+  the properties of your data that guarantees the uniqueness of the key you
+  have specified.
+
+  {{< warning >}}
+  If the key is not in fact unique, downstream consumers may not be able to
+  correctly interpret the data in the topic, and Kafka key compaction may
+  incorrectly garbage collect records from the topic.
+  {{< /warning >}}
+
+### Debezium
+
+<b>Syntax:</b> `ENVELOPE DEBEZIUM`
+
+The Debezium envelope wraps each event in an object containing a `before` and
+`after` field to indicate whether the event was an insertion, deletion, or
+update:
+
+```json
+// Insertion event.
+{"before": null, "after": {"field1": "val1", ...}}
+
+// Deletion event.
+{"before": {"field1": "val1", ...}, "after": null}
+
+// Update event.
+{"before": {"field1": "oldval1", ...}, "after": {"field1": "newval1", ...}}
 ```
 
-[//]: # "TODO(morsapaes) Add information about upsert key selection"
+Consider using the Debezium envelope if:
+
+  * You have downstream consumers that want update events to contain both the
```

Comment:
nit weird to have a single-element bulleted list

Reply:
Oh good point. I missed the second bullet that I meant to add here ("there is no natural key for the sink").